File: composeApp/src/commonMain/kotlin/com/checklistboteco/domain/security/PasswordPolicy.kt
package com.checklistboteco.domain.security

data class PasswordRuleResult(
    val message: String,
    val satisfied: Boolean
)

object PasswordPolicy {
    fun rules(password: String): List<PasswordRuleResult> {
        return listOf(
            PasswordRuleResult("Ao menos 6 caracteres", password.length >= 6),
            PasswordRuleResult("Ao menos uma letra maiúscula", password.any(Char::isUpperCase)),
            PasswordRuleResult("Ao menos um número", password.any(Char::isDigit)),
            PasswordRuleResult("Ao menos um caractere especial", password.any { !it.isLetterOrDigit() })
        )
    }

    fun isValid(password: String): Boolean = rules(password).all { it.satisfied }

    fun firstInvalidMessage(password: String): String? = rules(password).firstOrNull { !it.satisfied }?.message
}

File: Packages/Auth/Sources/Auth/PasswordPolicy.swift
import Foundation

struct PasswordRule: Identifiable, Equatable {
  let id: String
  let message: String
  let satisfied: Bool
}

enum PasswordPolicy {
  static func rules(for password: String) -> [PasswordRule] {
    [
      PasswordRule(id: "length", message: "Ao menos 6 caracteres", satisfied: password.count >= 6),
      PasswordRule(id: "uppercase", message: "Ao menos uma letra maiúscula", satisfied: password.contains { $0.isUppercase }),
      PasswordRule(id: "number", message: "Ao menos um número", satisfied: password.contains { $0.isNumber }),
      PasswordRule(id: "special", message: "Ao menos um caractere especial", satisfied: password.contains { !$0.isLetter && !$0.isNumber }),
    ]
  }

  static func isValid(_ password: String) -> Bool {
    rules(for: password).allSatisfy(\.satisfied)
  }
}
